server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1787,102 +1791,109 @@ private RepositoryException notStartedException() {
      *
      * @param listener listener to resolve with new repository data
      */
-    private void initializeRepoGenerationTracking(ActionListener<RepositoryData> listener) {
+    private boolean initializeRepoGenerationTracking(ActionListener<RepositoryData> listener) {
+        final SubscribableListener<RepositoryData> listenerToSubscribe;
+        final ActionListener<RepositoryData> listenerToComplete;
```

Comment:
NB previously we were potentially completing the listener under the mutex (e.g. by recursing back into `getRepositoryData`) - by introducing these local vars we can release the mutex before doing anything meaningful.